File: Test_for_unit/204_Count_Primes.cpp
/*############################################################################
#Author:polya 
#Emalil:polyaluthor@gmail.com
#204_Count_Primes.cpp
#Last updated:2019.10.22
#history version
#Version: 0.1
#Purpose:
#Main logic principle:
#Usage:
#     example: model.cpp
g++ model.cpp -o model -std=gnu++11
make model
./model
g++ 204_Count_Primes.cpp -o 204 -std=gnu++11
make 204
./204
############################################################################*/
//logic outline:print information
//include input and out library
//logic struture:
//
#include <iostream>
#include <ctime>
#include <vector>
#include <iterator>
#include <sstream>
#include <algorithm>
#include "stdio.h"
using namespace std;
//define class for solution
class Solution {
	public:
		int countPrimes(int num){
			if(num<=2) return 0;
			vector<bool> prime(num,true);
			prime[0]=false,prime[1]=false;
			for(int i=0;i<=sqrt(num);++i){
				if(prime[i]){
					for(int j=i*i;j<num;j+=i){
						prime[j]=false;
					}
				}
			}
			int sizes=prime.size();
			int nu=0;
			for(int k=0;k<sizes;k++){
				//printf("%d ",nu);
				if(prime[k]==1){
					printf("%d ",nu);
				//	cout<<nu<<endl;
				}
				nu=nu+1;
				//cout<<prime[k]<<endl;
			}
			return count(prime.begin(), prime.end(), true);
			}
	void displayInfo(vector<int>& nums,bool flag){
		if(flag==1){
			for (int i =0;i<nums.size();i++){
				cout<<nums[i]<<endl;
			}
		}else{
			// Convert vector to string
			ostringstream vts;
			vector<int> vec=nums;
			if (!vec.empty()){
			copy(vec.begin(), vec.end()-1,ostream_iterator<int>(vts, ", ")); 
			vts << vec.back();
			}
			cout<<vts.str()<<endl;
		}
	}
	void displayInfo(int nums){
		cout<<nums<<endl;
	}
};
//call main function
int main(int argc, char* argv[])
{
//self test ###################################################################
	clock_t begin = clock();
	cout<<"Hello polya,welcome to the programming world" <<endl;
	cout<<"===========================================" <<endl;
	cout<<"//out element: " <<endl;
//run programming############################
	int fPrime;
//	vector<int> f_redu;
	int num;
//arr={1,12,-5,-6,50,3};
	//arr={1,2,1,0,0,3,3,4};
	//arr={1,1,2};
	//arr={0,0,1,1,2,3,3,4};
	num=80;
	//unsigned int k=4;
	Solution f_prime;
//	fRedu.displayInfo(arr);
	fPrime=f_prime.countPrimes(num);
	cout<<" The out of ends:  "<<fPrime<<endl;
	f_prime.displayInfo(fPrime);
	cout<<" The plus One "<<endl;
//###########################################
	cout<<"===========================================" <<endl;
	clock_t end = clock();
	double elapsed_secs = static_cast<double>(end - begin) / CLOCKS_PER_SEC;
	time_t now = time(0);
	char* dt = ctime(&now);
	cout<<"The programming had finished at time : "<< dt << "==============================" <<endl;
	cout <<"The programming running time: "<< elapsed_secs << " s" << endl;
//out test ####################################################################
//unit test ###################################################################
	return 0;
}

File: Test_for_unit/204_Count_Primes.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
############################################################################
#Author:polya 
#Email:polyaluthor@gmail.com
#Last updated:2019.10.22
#Version: 1.0 first
#need do: unit test
#Purpose: 204_Count_Primes.py
#Main logic principle:
#
#
#Usage:
#     example: python 204_Count_Primes.py
############################################################################
"""
import os
import sys
import math
import datetime
import argparse
import pandas as pd
#
#
#Get Argv
#get ready parameters and out help inforamtion
#############################################################################
def get_args():
	parse=argparse.ArgumentParser(description='annot expression in one file')
	parse.add_argument('-u','--usage',help="usage: python $0  -in 1,2,3 ")
	parse.add_argument('-in','--in_list',type=str,help="input ready files ")
	parse.add_argument('-out','--out_file',type=str,help="set out name of files: eg.select.txt ")
	args=parse.parse_args()
	return vars(args)
############################################################################
#
#
#Do somthing for data
############################################################################
class Solution:
	def __init__(self,num):
		self.nums=num
		#print(self.k)
	def displayShow(self):
		#for i_iter in self.digits:
		#	print(i_iter)
		print("the input of number: "+str(self.nums))
	def countPrimes(self):
		if self.nums<=2:
			return 0
		prime=[1]*self.nums
		prime[0]=prime[1]=0
		for i in range(2,(int)(math.sqrt(self.nums)+1)):
			if prime[i]:
				for j in range(i+i,self.nums,i):
					prime[j]=0;
		#return  prime.count(1)
		count=0
		for i in range(0,len(prime)):
			if prime[i]:
				print(count)
			count=count+1
		return  prime.count(1)
#
#Call steps for ends
############################################################################
def call_main_steps():
	args=get_args()
	start =datetime.datetime.now()
	#arr=[1,12,-5,-6,50,3,100,20,11]
	#arr=[9,2,1,2,1,9]
	arr=[1,1,3,5,6,6,6.7]
	num=80
#self test##################################
	lena=Solution(num)
	ar_len=lena.displayShow()
	ar_len=lena.countPrimes()
	print("The count of prime of the number  : %s"%(ar_len))
	print("\n\n============================================================")
	print("Hello polya,welcome to the programming world")
	end=datetime.datetime.now()
	#print("The programming running time : "+str(end-start))
	print('The programming running time: %s Seconds'%(end-start))
	print("The programming finished time : "+str(end))
	print("============================================================")
#input from out############################
#unit test ################################ 
#
#main
############################################################################
if __name__=='__main__':
	call_main_steps()

